V2 docs (#93)
cohere-ai/cohere-developer-experience: fern/pages/v2/text-generation/chat-api.mdx

---
title: "Using the Chat API"
slug: "v2/docs/chat-api"

hidden: false 
description: "The document explains how to use the Chat API endpoint with Cohere LLMs to generate text responses in a conversational interface, including examples in Python, Java, and TypeScript. It also covers response structure, multi-turn conversations, and using a `conversation_id` to save chat history."
image: "../../../assets/images/4a5325a-cohere_meta_image.jpg"  
keywords: "Cohere, text generation, LLMs, generative AI"

createdAt: "Thu Feb 29 2024 18:05:29 GMT+0000 (Coordinated Universal Time)"
updatedAt: "Tue Jun 18 2024 07:20:15 GMT+0000 (Coordinated Universal Time)"
---

> The Chat API endpoint is used to generate text with Cohere LLMs. This endpoint facilitates a conversational interface, allowing users to send messages to the model and receive text responses.

Reviewer: we should mention that every message comes with "content" and an associated "role", which indicates who that message was sent from. Roles can be "user", "assistant", "system" and "tool".

<CodeBlocks>
```python PYTHON
import cohere
co = cohere.ClientV2(api_key="<YOUR API KEY>")

res = co.chat(
    model="command-r-plus",
    messages=[{"role": "user", "content": "Write a title for a blog post about API design. Only output the title text."}]
)

print(res.message.content[0].text) # "The Ultimate Guide to API Design: Best Practices for Building Robust and Scalable APIs"
```

> ```java JAVA

Author: Ok will remove those for now.

```
```java JAVA
public class ChatPost {
  public static void main(String[] args) {
    Cohere cohere = Cohere.builder().token("<YOUR API KEY>").build();

    NonStreamedChatResponse response = cohere.chat(
      ChatRequest.builder()
      	.model("command-r-plus-08-2024")
      	.message("Write a title for a blog post about API design. Only output the title text.")
    )

    System.out.println(response); // "The Art of API Design: Crafting Elegant and Powerful Interfaces"
  }
}
```
```typescript TYPESCRIPT
const { CohereClient } = require('cohere-ai');

> const cohere = new CohereClient({

Reviewer: @billytrend-cohere does this need to be CohereClientV2?

  token: '<YOUR API KEY>',
});

(async () => {
  const response = await cohere.chat({
    message: 'Write a title for a blog post about API design. Only output the title text.',
  });

  console.log(response.text)
})();
```
</CodeBlocks>

## Response Structure

Below is a sample response from the Chat API

```json JSON
{
  "id": "d656d46d-206b-4cc9-9baa-46ef8931cd18",
  "finish_reason": "COMPLETE",

>   "prompt": null,

Reviewer: ```suggestion
```

    "tool_plan": null,
    "content": [
      {
        "text": "The Ultimate Guide to API Design: Best Practices for Building Robust and Scalable APIs",
        "type": "text"
      }
    ],
    "citations": null,
    "role": "assistant"
  },
  "usage": null,
  "meta": {
    "api_version": {
      "version": "2",
      "is_experimental": true
    },
    "warnings": [
      "You are using an experimental version, for more information please refer to https://docs.cohere.com/versioning-reference"
    ],
    "billed_units": {
      "input_tokens": 17,
      "output_tokens": 16
    },
    "tokens": {
      "input_tokens": 215,
      "output_tokens": 16
    }
  }
}
```

Every response contains the following fields:

> - `message` the generated message from the model.

Author: should be remove non-python for now since only python SDK is available?

- `message` the generated message from the model.
- `id` the ID corresponding to this response.
- `finish_reason` can be one of the following:
  - `COMPLETE` the model successfully finished generating the message
  - `MAX_TOKENS` the model's context limit was reached before the generation could be completed
- `meta` contains information with token counts, billing etc.

## System Message

> A system message can be optionally provided in the request. A system message is a message provided to a model at the beginning of a conversation which dictates how the model should behave throughout.

Reviewer: i think we should write a better description for this. we should cover:

- developers can adjust the LLMs behavior by including a System message in the `messages` list
- those messages have the role set to `system`
- the system message contains instructions that the model will respect over any instructions sent in messages sent from other roles
- often used by developers to control the style in which the model communicates and to provide guidelines for how to handle various topics
- it is recommended to send the system message as the first element in the `messages` list

Author: Thanks for these suggestions. Will add.

```python PYTHON

> system_message = "You respond concisely, in about 5 words or less"

Reviewer: ```suggestion
import cohere
co = cohere.ClientV2("YOUR API KEY")

system_message = "You respond concisely, in about 5 words or less"
```

res = co.chat(
    model="command-r-plus",
    messages=[{"role": "system", "content": preamble},
              {"role": "user", "content": "Write a title for a blog post about API design. Only output the title text."}] # "Designing Perfect APIs"
)

print(res.message.content[0].text)
```

## Multi-Turn Conversations

> The user message in the Chat request can be sent together in the `messages` list to provide the model with conversational context:

Reviewer: ```suggestion
A single Chat request can encapsulate multiple turns of a conversation, where each message in the `messages` list appears in the order it was sent. Sending multiple messages can give the model context for generating a response.
```

```python PYTHON

> system_message = """## Task and Context

Reviewer: import statement + client initialization

> system_message = "You respond concisely, in about 5 words or less"

Reviewer: ```suggestion
import cohere
co = cohere.ClientV2("YOUR API KEY")

system_message = "You respond concisely, in about 5 words or less"
```
